File: Tutorials/VB/Tutorials/Tutorial14.vb
Imports NetOffice
Imports NetOffice.CollectionsGeneric
Imports NetOffice.Contribution.CollectionsGeneric
Imports Excel = NetOffice.ExcelApi

Public Class Tutorial14
    Implements ITutorial

    Dim _hostApplication As IHost

    Public Sub Run() Implements TutorialsBase.ITutorial.Run

        ' In some situations you want use NetOffice with an already running application.
        ' this tutorial shows how its possible.

        ' 1)
        '
        ' GetActiveInstance take the first instance in memory
        Dim application As Excel.Application = Excel.Application.GetActiveInstance()
        If Not IsNothing(application) Then
            application.Dispose()
        End If

        ' 2)
        '
        ' GetActiveInstances takes all instances in memory
        Dim applications As IDisposableSequence(Of Excel.Application) = Excel.Application.GetActiveInstances()
        applications.Dispose()

        ' 3)
        '
        ' Use special ctor to try access a running application first
        ' and if its failed create a new application
        application = New Excel.Application(New Core(), True)
        ' quit only if its a new application
        If Not application.FromProxyService Then
            application.Quit()
        End If
        application.Dispose()

        ' 4)
        '
        ' Creates instance from interop proxy
        Dim interopType As Type = Type.GetTypeFromProgID("Excel.Application")
        Dim proxy As Object = Activator.CreateInstance(interopType)
        application = COMObject.Create(Of Excel.Application)(proxy)
        application.Quit()
        application.Dispose()


        _hostApplication.ShowFinishDialog()

    End Sub

    Public ReadOnly Property Caption As String Implements TutorialsBase.ITutorial.Caption
        Get
            Return "Tutorial14"
        End Get
    End Property

    Public ReadOnly Property Description As String Implements TutorialsBase.ITutorial.Description
        Get
            Return "Accessing running applications"
        End Get
    End Property

    Public Sub Connect(ByVal hostApplication As TutorialsBase.IHost) Implements TutorialsBase.ITutorial.Connect

        _hostApplication = hostApplication

    End Sub

    Public Sub Disconnect() Implements TutorialsBase.ITutorial.Disconnect

    End Sub

    Public ReadOnly Property Panel As System.Windows.Forms.UserControl Implements TutorialsBase.ITutorial.Panel
        Get
            Return Nothing
        End Get
    End Property

    Public ReadOnly Property Uri As String Implements TutorialsBase.ITutorial.Uri
        Get
            Return FormMain.DocumentationBase & "Tutorial14_EN_VB.html"
        End Get
    End Property

End Class

File: Tutorials/C#/Tutorials/Tutorial14.cs
using System;
using System.Windows.Forms;
using TutorialsBase;
using NetOffice;
using Excel = NetOffice.ExcelApi;
using NetOffice.Running;

namespace TutorialsCS4
{
    public class Tutorial14 : ITutorial
    {
        public void Run()
        {
            // In some situations you want use NetOffice with an already running application.
            // this tutorial shows how its possible.

            // 1)
            //
            // GetActiveInstance take the first instance in memory
            Excel.Application application = ProxyService.GetActiveInstance<Excel.Application>();
            if (null != application)
                application.Dispose();

            // 2)
            //
            // GetActiveInstances takes all instances in memory
            var applications = ProxyService.GetActiveInstances<Excel.Application>();
            applications.Dispose();

            // 3)
            //
            // Use special ctor to try access a running application first
            // and if its failed create a new application
            application = new Excel.ApplicationClass(new Core(), true);
            // quit only if its a new application
            if (!application.FromProxyService)
                application.Quit();
            application.Dispose();

            // 4)
            //
            // Creates instance from interop proxy
            Type interopType = Type.GetTypeFromProgID("Excel.Application");
            object proxy = Activator.CreateInstance(interopType);
            application = COMObject.Create<Excel.Application>(proxy);
            application.Quit();
            application.Dispose();


            HostApplication.ShowFinishDialog();
        }

        public void Connect(IHost hostApplication)
        {
            HostApplication = hostApplication;
        }

        public void Disconnect()
        {

        }

        public string Uri
        {
            get { return Program.DocumentationBase + "Tutorial14_EN_CS.html"; }
        }

        public string Caption
        {
            get { return "Tutorial14"; }
        }

        public string Description
        {
            get { return "Accessing running applications"; }
        }

        public UserControl Panel
        {
            get { return null; }
        }

        internal IHost HostApplication { get; private set; }
    }
}
